### ISSUE TITLE
GotoTarget lifetime

### ISSUE THREAD

**Issue by int19h** [Microsoft/debug-adapter-protocol#39]
The specification does not define the lifetime of a `GotoTarget` as obtained from a given `GotoTargetsRequest`, so it is unclear if it's legal for an IDE to cache them and reuse them later (possibly across `goto`, across `continue`, or in different frames that just happen to use the same source), or whether the adapter can clean up associated resources more aggressively - e.g. on invalidate them on every `goto`.

It's also unclear whether the adapter can return different IDs for the same file/line on a subsequent request, even if it ensures that old IDs are still valid to use in `goto` - and if it can do so, then how should the IDE interpret this.

**Comment by gregg-miskelly** [comment]
CC @andrewcrawley 

My suggested documentation: Ids are valid until the next continue. If multiple `GotoTargetsRequest` are issued by a client, the debug adapter is allowed, but not required, to return the same id if the underlying execution location is the same between the two requests.

**Comment by Trass3r** [comment]
Could someone explain a use-case for  multiple GotoTargetsRequests?
VSCode just issues 1 targets request and directly feeds the id into the goto request for its Jump to Cursor functionality, ignoring anything else inside the GotoTarget: https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/src/vs/workbench/contrib/debug/browser/debugCommands.ts#L169-183
I can barely come up with an example for returning more than 1 target (template functions come to mind but why would it be ambiguous if you are already in the same function) let alone multiple target requests.
Also I don't know what the other GotoTarget fields could be useful for (line?).

**Comment by gregg-miskelly** [comment]
I am not sure why a client might want to issue multiple `gotoTargets` requests, but it also seems wrong not to handle them -- the DAP uses handles in lots of other places, so I don't know why an adapter would do things differently here.

`GotoTargetsArguments` doesn't specify a thread id, so I don't think an adapter could actually know which template instantiation to return.

[`line` in `GotoTarget`](https://github.com/microsoft/debug-adapter-protocol/blob/4b2eba24dd77f3bcbb3400596358d5cfe737446a/debugAdapterProtocol.json#L3552) is useful for the same reason as in breakpoints - the incoming line is where the user requested. But the debug adapter may find that it doesn't have an executable code associated with that line, but it does have executable code associated with lines in close proximity. An IDE could then choose to use that line info in a disambiguation dialog.

FYI: If you are asking about this because you are thinking about implementing this in MIEngine, `GotoTargets` needs to call [IDebugProgram2.EnumCodeContexts](https://github.com/microsoft/MIEngine/blob/e1b8dda3c7265b3c3a771a46fe283c36b498d169/src/MIDebugEngine/AD7.Impl/AD7Engine.cs#L843).

**Comment by Trass3r** [comment]
> I am not sure why a client might want to issue multiple `gotoTargets` requests, but it also seems wrong not to handle them -- the DAP uses handles in lots of other places, so I don't know why an adapter would do things differently here.

I guess confusion about the spec like in this ticket, vscode as the reference implementation does it as described above and it makes the implementation much easier (at least if you assume IDs need to be globally unique, that crucial 'certain IDs are valid only until the next continue' information is buried in some tickets on Github).

> `GotoTargetsArguments` doesn't specify a thread id, so I don't think an adapter could actually know which template instantiation to return.

Interesting. Any other use-case for multiple locations?

> [`line` in `GotoTarget`](https://github.com/microsoft/debug-adapter-protocol/blob/4b2eba24dd77f3bcbb3400596358d5cfe737446a/debugAdapterProtocol.json#L3552) is useful for the same reason as in breakpoints - the incoming line is where the user requested. But the debug adapter may find that it doesn't have an executable code associated with that line, but it does have executable code associated with lines in close proximity.

I see, should be added to the docs.

**Comment by gregg-miskelly** [comment]
> Any other use-case for multiple locations?

For compiled languages, I think of the `gotoTargets` request as akin to asking: "if I was to set a breakpoint on this line, where would the breakpoint mind?" without actually setting the breakpoint. So you could get multiple answers if:
* The function was compiled multiple times (ex: C++ templates)
* The line was mapped to multiple instruction groups (ex: optimizations, compiler bugs, `for` loops in C-like languages, etc)
* Potentially a managed runtime might JIT a method multiple times (ex: fast JIT and optimizing JIT) and expose that

**Comment by Trass3r** [comment]
Yeah it's easy to handle breakpoints as for example gdb just sets 1 per associated address. But `jump` will probably fail if it's ambiguous.

**Comment by Trass3r** [comment]
> I am not sure why a client might want to issue multiple `gotoTargets` requests

If you want to implement jumping by dragging the current line marker like in VS you'd probably have to do that.
